lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -0,0 +1,517 @@
+# to 0 if you are raising the major API version
+LIBPATCH = 12
+
+
+class LogAdapter(logging.LoggerAdapter):
+    """Adapter for the logger to prepend a prefix to all log lines."""
+
+    prefix = "vault_tls"
+
+    def process(self, msg, kwargs):
+        """Decides the format for the prepended text."""
+        return f"[{self.prefix}] {msg}", kwargs
+
+
+logger = LogAdapter(logging.getLogger(__name__), {})
+
+SEND_CA_CERT_RELATION_NAME = "send-ca-cert"
+TLS_CERTIFICATE_ACCESS_RELATION_NAME = "tls-certificates-access"
+CA_CERTIFICATE_JUJU_SECRET_LABEL = "self-signed-vault-ca-certificate"
+
+VAULT_CA_SUBJECT = "Vault self signed CA"
+
+
+class TLSMode(Enum):
+    """This class defines the different modes of TLS configuration.
+
+    SELF_SIGNED: The charm will generate a self signed certificate.
+    TLS_INTEGRATION: The charm will use the TLS integration relation.
+    """
+
+    SELF_SIGNED = 1
+    TLS_INTEGRATION = 2
+
+
+# TODO Move this class, it doesn't belong here.
+class WorkloadBase(ABC):
+    """Define an interface for the Machine and Container classes."""
+
+    @abstractmethod
+    def exists(self, path: str) -> bool:
+        """Check if a file exists in the workload."""
+        pass
+
+    @abstractmethod
+    def pull(self, path: str) -> TextIO:
+        """Read file from the workload."""
+        pass
+
+    @abstractmethod
+    def push(self, path: str, source: str) -> None:
+        """Write file to the workload."""
+        pass
+
+    @abstractmethod
+    def make_dir(self, path: str) -> None:
+        """Create directory in the workload."""
+        pass
+
+    @abstractmethod
+    def remove_path(self, path: str, recursive: bool = False) -> None:
+        """Remove file or directory from the workload."""
+        pass
+
+    @abstractmethod
+    def send_signal(self, signal: int, process: str) -> None:
+        """Send a signal to a process in the workload."""
+        pass
+
+    @abstractmethod
+    def restart(self, process: str) -> None:
+        """Restart the workload service."""
+
+    @abstractmethod
+    def stop(self, process: str) -> None:
+        """Stop a service in the workload."""
+        pass
+
+
+class VaultCertsError(Exception):
+    """Exception raised when a vault certificate is not found."""
+
+    def __init__(self, message: str = "Could not retrieve vault certificates from local storage"):
+        self.message = message
+        super().__init__(self.message)
+
+
+class File(Enum):
+    """This enum determines which files are expected of the library to read."""
+
+    CERT = auto()
+    KEY = auto()
+    CA = auto()
+    AUTOUNSEAL_CA = auto()
+
+
+class VaultTLSManager(Object):
+    """This class configures the certificates within Vault."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        service_name: str,
+        tls_directory_path: str,
+        workload: WorkloadBase,
+        common_name: str,
+        sans_dns: FrozenSet[str] = frozenset(),
```

Comment:
You'll probably want to validate this with pydantic types

Reply:
I'm not sure I see why, could you please help

Reply:
To make sure the value is a valid IP

Reply:
I agree some work is needed for data validation in most of our charms and libraries. I prefer to keep that out of the context of this PR.